pkg/local-gadget-manager/local-gadget-manager.go

```diff
@@ -0,0 +1,409 @@
+	traceFactories map[string]gadgets.TraceFactory
+
+	// tracers by name
+	tracers map[string]tracer
+
+	// containers by id
+	containers map[string]pb.ContainerDefinition
+
+	// subs contains a list of subscribers
+	pubsub *pubsub.GadgetPubSub
+
+	runcNotifier *runcfanotify.RuncNotifier
+
+	dockercli *client.Client
+}
+
+type tracer struct {
+	gadget        string
+	name          string
+	factory       gadgets.TraceFactory
+	traceResource *gadgetv1alpha1.Trace
+	gadgetStream  *stream.GadgetStream
+}
+
+func (l *LocalGadgetManager) ListGadgets() []string {
+	gadgets := []string{}
+	for name := range l.traceFactories {
+		gadgets = append(gadgets, name)
+	}
+	sort.Strings(gadgets)
+	return gadgets
+}
+
+func (l *LocalGadgetManager) ListOperations(gadget string) []string {
+	return []string{"start", "stop", "generate"}
+}
+
+func (l *LocalGadgetManager) ListTraces() []string {
+	traces := []string{}
+	for name := range l.tracers {
+		traces = append(traces, name)
+	}
+	sort.Strings(traces)
+	return traces
+}
+
+func (l *LocalGadgetManager) ListContainers() []string {
+	containers := []string{}
+	for id := range l.containers {
+		containers = append(containers, id)
+	}
+	sort.Strings(containers)
+	return containers
+}
+
+func (l *LocalGadgetManager) AddTracer(gadget, name, containerFilter string) error {
+	factory, ok := l.traceFactories[gadget]
+	if !ok {
+		return fmt.Errorf("unknown gadget")
+	}
+	_, ok = l.tracers[name]
+	if ok {
+		fmt.Printf("Error: trace already exists")
+		return nil
+	}
+
+	traceResource := &gadgetv1alpha1.Trace{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: "gadget",
+		},
+		Spec: gadgetv1alpha1.TraceSpec{
+			Node:   "local",
+			Gadget: gadget,
+			Filter: &gadgetv1alpha1.ContainerFilter{
+				Namespace:     "local",
+				Podname:       containerFilter,
+				ContainerName: "",
+				Labels:        map[string]string{},
+			},
+			RunMode:    "Manual",
+			OutputMode: "Stream",
+		},
+	}
+
+	l.tracers[name] = tracer{
+		gadget:        gadget,
+		name:          name,
+		factory:       factory,
+		traceResource: traceResource,
+		gadgetStream:  stream.NewGadgetStream(),
+	}
+	return nil
+}
+
+func (l *LocalGadgetManager) Operation(name, opname string) error {
+	tracer, ok := l.tracers[name]
+	if !ok {
+		return fmt.Errorf("unknown trace")
+	}
+
+	if opname != "" {
+		gadgetOperation, ok := tracer.factory.Operations()[opname]
+		if !ok {
+			return fmt.Errorf("Unknown operation %q", opname)
+		}
+		tracerNamespacedName := tracer.traceResource.ObjectMeta.Namespace +
+			"/" + tracer.traceResource.ObjectMeta.Name
+		gadgetOperation.Operation(tracerNamespacedName, tracer.traceResource)
+	}
+
+	fmt.Printf("State: %s\n", tracer.traceResource.Status.State)
+	fmt.Printf("OperationError: %s\n", tracer.traceResource.Status.OperationError)
+	fmt.Printf("Output: %s\n", tracer.traceResource.Status.Output)
+
+	return nil
+}
+
+func (l *LocalGadgetManager) Show(name string) error {
+	return l.Operation(name, "")
+}
```

Comment:
What about having a `func (l *LocalGadgetManager) show(tracer ...)` that is used by both, `Operation()` and `Show()`. By doing that you avoid handling `""` as a special case.